adr/0003-multiple-templates-per-component.md

```diff
@@ -0,0 +1,52 @@
+# 1. Allow multiple templates 
+
+Date: 2021-08-30
+
+## Status
+
+Proposed.
+
+## Context
+
+As components become larger (for example, because you are implementing a whole page), it becomes
+useful to be able to extract sections of the view to a different file. ActionView has
+partials, and ViewComponent lacks a similar mechanism. 
+
+ActionView partials have the problem that their interface is not introspectable. Data
+may be passed into the partial via ivars or locals, and it is impossible to know
+which without actually opening up the file. Additionally, partials are globally 
+invocable, thus making it difficult to detect if a given partial is in use or not,
+and who are its users.
+
+An option would be to extract another Component for the extracted section, but that
+has the following drawbacks:
+
+1. It creates a new public class.
+2. If new component is invoked repeatedly (eg, for a list of items), this creates
+   GC pressure by creating lots of intermediate objects.
+
+## Decision
+
+We will allow having multiple templates in the sidecar asset. Each asset will be compiled to
```

Comment:
I've read this a number of times, and my brain keeps short-circuiting to "this is what slots are for." Slots take arguments and are invoked in a specific place in the template. To me, they have identical behavior to what you're proposing 🤔 

I do see your point about wanting to render sub-templates without having to turn them into a full component though. What about something like this:

```ruby
# app/components/my_component.rb

class MyComponent < ViewComponent::Base
  # notice we're not passing any component class, string, or lambda here
  renders_many :posts
end
```

```erb
<%# app/components/my_component.html.erb %>
<%= render(MyComponent.new) do |c| %>
  <%# renders_many implicitly makes all these kwargs available to the sub template. %>
  <%# The sub template's filename matches the singular name of the slot. %>
  <% c.post(post: p) %>
<% end %>
```

```erb
<%# app/components/my_component/_post.html.erb %>
<%= post.title %>
```

However, even something like this makes me a little uncomfortable. The idea behind view components is to use, well, _components_. In my view, you might as well use regular 'ol actionview if you're not breaking view code up into components 😄

Reply:
> I've read this a number of times, and my brain keeps short-circuiting to "this is what slots are for." Slots take arguments and are invoked in a specific place in the template. To me, they have identical behavior to what you're proposing thinking

No, slots are a primarily a way to *inject* content into components (eg, the message in an Alert component). This is all about internal organization of the component.

> I do see your point about wanting to render sub-templates without having to turn them into a full component though. What about something like this:
> <snip code>

I don't understand the code. Where do the posts come from and where is the loop? Is MyComponent recursiveness intended?

One comment though: my proposal avoids the magical variables you pass via kwargs. Your proposal makes the interface between the partial and the main view invisible (ie, you would need to go read the view, and the main component, to be able to know which variables I need to pass). My proposal makes such variables explicit, and @joelhawksley's proposal improves on that to allow all ruby argument forms.

> The idea behind view components is to use, well, components. In my view, you might as well use regular 'ol actionview if you're not breaking view code up into components 

Well, not all components are the same size. If the component is relatively large, splitting out sub-methods is a reasonable thing to do. All this merge does is allow you to write that method in ERB. Reasonable people might disagree, but I don't regard `RowComponent` a component: its coupling with the `TableComponent` is so tight that they are effectively a large super-component. I'd prefer to not have the row be a public interface.

Reply:
> > No, slots are a primarily a way to inject content into components (eg, the message in an Alert component). This proposal is all about internal organization of the component.
> 
> I see, so you're saying part of the benefit of rendering sidecar templates is that it allows you to organize the component's constituent pieces and keep them "private," i.e. without opening up a slot? That's a really interesting idea I hadn't considered.

Exactly! Please help me retitle/re-describe the issue, apparently it is not clear.

Do note that what I want is not exactly "private" as is understood in other languages. I very much want derived components to be able to override the erb files, if they need to.  See use case 2 here: https://github.com/github/view_component/issues/387#issuecomment-663785026

> 
> > I don't understand the code. Where do the posts come from and where is the loop? Is MyComponent recursiveness intended?
> 
> Ah sorry, the recursiveness is not intended, I mistakenly rendered `MyComponent` in the wrong file. I have updated the comment and added the missing loop.

Thanks. I still think this PR would improve things, since it would simplify the `MyComponent` interface to:

```erb
<%= render(MyComponent.new(posts: Post.all)) %>
```

> 
> > Your proposal makes the interface between the partial and the main view invisible (ie, you would need to go read the view, and the main component, to be able to know which variables I need to pass). My proposal makes such variables explicit.
> 
> That's definitely a valid concern. Could be mitigated by passing a lambda and rendering the ERB inside it, eg:
> 
> ```ruby
> # app/components/my_component.rb
> 
> class MyComponent < ViewComponent::Base
>   renders_many :posts, lambda do |p:|
>     render_erb '_sub_component', post: p
>   end
> end
> ```

But then you need something like this PR that allows extra ERB files :wink: 

I'm open to alternative interfaces. As I indicated on https://github.com/github/view_component/pull/451#issuecomment-681172299, I agree the current interface is very ugly, and we should strive for a better one. 

> 
> > All this merge does is allow you to write that method in ERB.
> 
> I do think it's potentially a good idea to allow rendering additional, component-scoped sidecar assets.

:smile:

Reply:
> If the component is relatively large, splitting out sub-methods is a reasonable thing to do. All this merge does is allow you to write that method in ERB. Reasonable people might disagree, but I don't regard `RowComponent` a component: its coupling with the `TableComponent` is so tight that they are effectively a large super-component. I'd prefer to not have the row be a public interface.

One of the biggest benefits of adopting components is that you can look at a minimal number of files and understand the component's behavior and its responsibility (components are very focused, by design). When thinking in terms of components, if a row has its own behavior it's a good sign that it should be its own component. I can understand the desire to keep it coupled to Table and not extract a component, but that  introduces other concerns like the table component losing focus, doing too much, as well as some testing concerns. It's not enforced, but it could always be undocumented or left as a private API though documentation/comments.

> > I see, so you're saying part of the benefit of rendering sidecar templates is that it allows you to organize the component's constituent pieces and keep them "private," i.e. without opening up a slot? That's a really interesting idea I hadn't considered.
> 
> Exactly! Please help me retitle/re-describe the issue, apparently it is not clear.

I think I understand the proposal, but I'm not sure I see the benefits of extracting separate ERB files. Similar to the above, I think it allows components to be broken apart in ways that I'm concerned would allow components to grow too large or take on behaviors that would be better encapsulated in separate object(s)/component(s). You also lose the benefit of being able to open two files in your editor and see the entire behavior of the component.

If we do accept this proposal, we'll end up with three different ways to render content in the framework, as well as the methods provided by Rails, like partials.

* Components (encapsulates behavior)
* Slots (composes components)
* Component partials (renders more HTML in the scope of a component?)

I think it would help to come up with a name for the abstraction that describes it well. I also have some other questions, like does it still go through `render`, are we bypassing Rails completely, is there parts of Rails we can take advantage of etc.?

That's a lot, but I did have one more question. Does any other component oriented framework have prior art we could take inspiration from here? I'm interested in seeing this behavior elsewhere and how it works and fits in a component based architecture.

Reply:
> > If the component is relatively large, splitting out sub-methods is a reasonable thing to do. All this merge does is allow you to write that method in ERB. Reasonable people might disagree, but I don't regard `RowComponent` a component: its coupling with the `TableComponent` is so tight that they are effectively a large super-component. I'd prefer to not have the row be a public interface.
> 
> One of the biggest benefits of adopting components is that you can look at a minimal number of files and understand the component's behavior and its responsibility (components are very focused, by design). 

I'm not very convinced by this. The way I see it, adopting components introduces encapsulation into the previous wild west of views. There's nothing that says a given component is small or large. The entire view could be a component!

> When thinking in terms of components, if a row has its own behavior it's a good sign that it should be its own component. I can understand the desire to keep it coupled to Table and not extract a component, but that introduces other concerns like the table component losing focus, doing too much, as well as some testing concerns. It's not enforced, but it could always be undocumented or left as a private API though documentation/comments.

Introducing generic-in-name-only concepts hurts readability IMO. If I'm designing the `UserProfileComponent`, do I really need `UserProfilePictureComponent`? Will it be used anywhere else? If the answer is no, I believe it should just be part of the `UserProfileComponent`. Otherwise, you end up navigating a very deep tree of components.

Plus, in this example now you have to look at 4 files instead of 3 :wink: 

> 
> > > I see, so you're saying part of the benefit of rendering sidecar templates is that it allows you to organize the component's constituent pieces and keep them "private," i.e. without opening up a slot? That's a really interesting idea I hadn't considered.
> > 
> > 
> > Exactly! Please help me retitle/re-describe the issue, apparently it is not clear.
> 
> I think I understand the proposal, but I'm not sure I see the benefits of extracting separate ERB files. Similar to the above, I think it allows components to be broken apart in ways that I'm concerned would allow components to grow too large or take on behaviors that would be better encapsulated in separate object(s)/component(s). You also lose the benefit of being able to open two files in your editor and see the entire behavior of the component.

I encourage you to think of this the other way around. This PR is only needed because we have no way to define multiple "methods" in ERB. If ERB had a native way to pack multiple templates into a single file, would you have ViewComponent actively block that ability? Because ERB does not have that facility, this is the second best. I think inventing yet another file format to pack multiple templates is a bad idea.

> 
> If we do accept this proposal, we'll end up with three different ways to render content in the framework, as well as the methods provided by Rails, like partials.
> 
> * Components (encapsulates behavior)
> * Slots (composes components)
> * Component partials (renders more HTML in the scope of a component?)

Do note that component partials are not meant to be externally available. Perhaps I can make that explicit by marking them `private` or `protected`?

> 
> I think it would help to come up with a name for the abstraction that describes it well. I also have some other questions, like does it still go through `render`, are we bypassing Rails completely, is there parts of Rails we can take advantage of etc.?

I'm not sure what you are aiming at here, but this does not use `render`. It should be a pure function taking some inputs and returning some html.

> That's a lot, but I did have one more question. Does any other component oriented framework have prior art we could take inspiration from here? I'm interested in seeing this behavior elsewhere and how it works and fits in a component based architecture.

I'm not sure it counts, but React/JSX allows creating arbitrary html wherever you want, so you can have methods returning sub-components:

```jsx
class Test extends React.Component {
  render() {
    return <h1>Hello, {this.someFunc()}</h1>;
  }
  someFunc() {
    return <small>{this.props.name}</small>;
  }
}
ReactDOM.render(<Test name="Felipe" />, document.getElementById('root'));
```

This would print

```html
<h1>Hello, <small>Felipe</small></h1>
```

Reply:
> Does any other component oriented framework have prior art we could take inspiration from here? I'm interested in seeing this behavior elsewhere and how it works and fits in a component based architecture.

Komponent allows [component partials](https://github.com/komposable/komponent#component-partials)

Reply:
> I'm not sure it counts, but React/JSX allows creating arbitrary html wherever you want, so you can have methods returning sub-components:

That's possible today in ViewComponent via `content_tag` right? The only thing that's missing is a way to render erb. Not saying it shouldn't, but it is possible to implement that functionality today without it being a library feature. The `_after_compile` hook would allow some optimizations too, I believe.

> Komponent allows [component partials](https://github.com/komposable/komponent#component-partials) (but the example in the documentation is overly simplistic)

Thanks! I'll take a peek at that. I'm curious, did you utilize component partials often?

Reply:
@joelhawksley and I talked about this during our 1:1 today, and we came to the conclusion that there are really two feature requests here: 1) the ability to render arbitrary templates from slots, and 2) private slots.

As I said earlier, slots are the mechanism the view_component library has for composing a component from constituent parts. I think a lot of us maintainers are very hesitant to add another way to render templates because of the confusion we already see around slots, and the complexity of the approaches we've discussed here in this issue feel fairly prohibitive. In our 1:1, Joel and I tried to come up with a solution based on the one he proposed in an earlier comment and realized after ~30 minutes that we were just reimplementing actionview 😅  

The alternative we settled on was to be able to specify a template path in the slot definition. This is pretty similar to my last comment, although makes the template explicit. The other necessary feature is private slots. A private slot would only be usable by templates rendered "inside" the component. Trying to use them outside the component would raise an error.

Let me try to rewrite your table component using this technique.

```erb
<%# app/views/foo/index.html.erb %>
<%= render(TableComponent.new(rows: @rows)) do |component| %>
  <%# this raises an error %>
  <% c.column %>
<% end %>
```

```ruby
# app/components/table_component.rb
class TableComponent < ApplicationComponent
  renders_many :columns, template: 'column.html.erb', public: false

  def initialize(rows:)
    @rows = rows
  end
end
```

```erb
<%# app/components/table_component.html.erb %>
<table>
  <% @rows.each do |row| %>
    <tr>
      <% column(row: row, attribute: :name) %>
   </tr>
  <% end %>
</table>
```

```erb
<%# app/components/table_component/column.html.erb %>
<td><%= row.send(attribute) %></td>
```

There are a number of nice benefits to this approach:

1. Leverages the existing slots API.
1. Does not require instantiating components, so should be relatively performant.
1. Does what you want (I think?)

Reply:
I think the example does not fit my problem. In your example, the user specifies the columns (good!). But what I wanted was to extract the `row`, and not expose that to the user. How would I achieve that with this proposal?

Additionally, there is no specifying of the arguments, which:
1. Makes the API implicit rather than explicit
2.  We would need to invoke the template via `instance_exec` / `binding`, and that is slower.

I do like that there are fewer pieces though.

Reply:
Maybe we could pair to brainstorm how this could end up looking like?

Reply:
Hmm I see what you mean. What about this?

```ruby
# app/components/table_component.rb
class TableComponent < ApplicationComponent
  renders_many :rows, template: 'row.html.erb', public: false
  renders_many :columns, template: 'column.html.erb', public: false

  def initialize(rows:)
    @rows = rows
  end
end
```

```erb
<%# app/components/table_component.html.erb %>
<table>
  <% @rows.each do |row| %>
    <% row(row: row) %>
  <% end %>
</table>
```

```erb
<%# app/components/row.html.erb %>
<tr>
  <% column(row: row, attribute: :name) %>
</tr>
```

```erb
<%# app/components/table_component/column.html.erb %>
<td><%= row.send(attribute) %></td>
```

Although now that I write this out, it feels a little weird to have rows and columns at the same level.

> 1. Makes the API implicit rather than explicit

That's true, although I don't know if that's avoidable given that we're actively trying to remove `RowComponent` and `ColumnComponent`, which would have provided it.

> 2. We would need to invoke the template via `instance_exec` / `binding`, and that is slower.

Unless I'm misunderstanding what you mean, there should be no need for either of these. Slots are already implemented by dynamically defining methods on the component class. We could do the same thing with these private slots that render additional templates.

@fsateler I want to thank you for your patience as we try to figure this out. This issue has prompted a huge amount of discussion internally! I think it would make sense for all of us to hop on a call at some point to brainstorm. Feel free to DM me [on Twitter](https://twitter.com/camertron).

Reply:
> Although now that I write this out, it feels a little weird to have rows and columns at the same level.

Indeed, but in your example, `columns` should be the public slot for the user:

```erb
<%= render TableComponent.new(rows: @rows) do |c| %>
  <%# we want a table with name and age columns %>
  <% c.column(:name) %>
  <% c.column(:age) %>
<% end %>
```

> > 2. We would need to invoke the template via instance_exec / binding, and that is slower.
> Unless I'm misunderstanding what you mean, there should be no need for either of these.

Point 2 is a consequence of point 1. You need to somehow make the dynamically passed variables (in this example, `row` and `attribute`) as available in the invoked method binding. Where would be the canonical list to define the method parameters? (This is what the original PR tries to solve with the `template_arguments` method).

Reply:
> I'm curious, did you utilize component partials often?

@BlakeWilliams it's used 3 times in the largest codebase I work on (about ~100 components). It's mainly used to extract part of the template to another file, when the component template grows very big, but no part of the template would make sense as a reusable separate component.